src/hash.c

```diff
@@ -46,19 +46,41 @@ static uint32_t hash_double(double x) {
   return hash_int64(value.i);
 }
 
-static uint32_t hash_char(SEXP x) {
+// `hash_char_sexp()` uses the fact that R has a global string pool, and just
+// hashes the pointer into that string pool. This is very fast, but has the
+// side effect of not being reproducible from one R session to the next.
+// `hash_char()` is slower, but hashes the underlying char array, and is
+// reproducible.
+
+static uint32_t hash_char_sexp(SEXP x) {
   return hash_int64((intptr_t) x);
 }
 
+// djb2 - http://www.cse.yorku.ca/~oz/hash.html
+static uint32_t hash_char(SEXP x) {
+  uint32_t hash = 5381;
+
+  const char *string = CHAR(x);
```

Comment:
What happens with NAs here?

Reply:
It seems to match `"NA"`

``` r
library(vctrs)

vec_hash(NA_character_, pool = FALSE)
#> [1] 4d ee 90 9e

vec_hash("NA", pool = FALSE)
#> [1] 4d ee 90 9e
```

<sup>Created on 2019-11-13 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9000)</sup>

https://github.com/wch/r-source/blob/8019f9819095f98d4069459aa787b0d222a248e2/src/main/names.c#L1189

Reply:
It looks like `digest()` get's around this when they `serialize()` input, which they do by default. But their results are identical if you don't serialize.

``` r
library(digest)

# Look different by default
digest(NA_character_)
#> [1] "55611e71b358a30158c61810ad802435"
digest("NA")
#> [1] "279f2abfb74bae808fc0ff984b295c33"

# Without serializing they look the same
digest(NA_character_, serialize = FALSE)
#> [1] "d4cd0dabcf4caa22ad92fab40844c786"
digest("NA", serialize = FALSE)
#> [1] "d4cd0dabcf4caa22ad92fab40844c786"

# Different because of the serialization
serialize(NA_character_, connection = NULL, version = 2L)
#>  [1] 58 0a 00 00 00 02 00 03 06 00 00 02 03 00 00 00 00 10 00 00 00 01 00
#> [24] 00 00 09 ff ff ff ff
serialize("NA", connection = NULL, version = 2L)
#>  [1] 58 0a 00 00 00 02 00 03 06 00 00 02 03 00 00 00 00 10 00 00 00 01 00
#> [24] 04 00 09 00 00 00 02 4e 41
```

Reply:
I wonder if this would work. The idea would be that normally when we convert an individual `char` into an `int` (on the line `int chr = *string++`), that `chr` integer value is a value between 0-127 on the ASCII table. So if we use `-1` for `NA` maybe that would be enough?

https://www.cs.cmu.edu/~pattis/15-1XX/common/handouts/ascii.html

```
static uint32_t hash_char(SEXP x) {
  uint32_t hash = 5381;

  if (x == NA_STRING) {
    return ((hash << 5) + hash) - 1;
  }

  const char *string = CHAR(x);
  int chr = *string++;

  while (chr) {
    hash = ((hash << 5) + hash) + chr; /* hash * 33 + chr */
    chr = *string++;
  }

  return hash;
}
```

Reply:
What about the empty string `""`? IIUC `obj_hash("")` and `obj_hash(c("", ""))` will both be `0` currently, when bool/stable is TRUE.

Reply:
I'm not sure that `vec_hash(NA) == vec_hash("NA")` is a problem. And if `obj_hash("") == 0` is a problem then we should just run it through our hash scramble.

Reply:
I'm not sure I understand the empty string problem

``` r
vctrs:::obj_hash("", pool = FALSE)
#> [1] 77 08 6f 3c

vctrs:::obj_hash(c("", ""), pool = FALSE)
#> [1] 0c 29 d6 6b
```

Reply:
It seems there is none.